src/data/finset/noncomm_prod.lean

```diff
@@ -27,41 +27,39 @@ the two must be equal.
 
 -/
 
-variables {α β γ : Type*} (f : α → β → β) (op : α → α → α)
+variables {F ι α β γ : Type*} (f : α → β → β) (op : α → α → α)
 
 namespace multiset
 
 /-- Fold of a `s : multiset α` with `f : α → β → β`, given a proof that `left_commutative f`
 on all elements `x ∈ s`. -/
-def noncomm_foldr (s : multiset α) (comm : ∀ (x ∈ s) (y ∈ s) b, f x (f y b) = f y (f x b)) (b : β) :
+def noncomm_foldr (s : multiset α)
+  (comm : {x | x ∈ s}.pairwise $ λ x y, ∀ b, f x (f y b) = f y (f x b)) (b : β) :
```

Comment:
I just found we also have [multiset.pairwise](https://leanprover-community.github.io/mathlib_docs/data/multiset/basic.html#multiset.pairwise). Should we use it throughout the file? It may not be convenient to work with directly since is defined as "there exists a list such that pairwise ...", but could be if we build enough APIs around it (e.g. that `s.pairwise r` iff `{x | x ∈ s}.pairwise r` if `r` is symmetric).

Reply:
It seems `multiset.pairwise` doesn't have much API and that probably indicate that it isn't used much in mathlib; also the use of "exists" seems weird and make it not well-behaved; with `forall` it would be better (no need of symmetry of `r` for the equivalence of the three above). So I think I am satisfied with your current spelling using `set.pairwise`.

Reply:
Yeah, if only for compatibility with the other `noncomm_prod`, it seems better to stick to `set.pairwise`.

Reply:
Yeah, notice that [multiset.nodup](https://leanprover-community.github.io/mathlib_docs/data/multiset/nodup.html#multiset.nodup) also doesn't use multiset.pairwise but instead directly lifts list.nodup (which works because `ne` is symmetric).

Reply:
FYI: we do have the equivalence between multiset.pairwise and list.pairwise ([multiset.pairwise_coe_iff_pairwise](https://leanprover-community.github.io/mathlib_docs/data/multiset/basic.html#multiset.pairwise_coe_iff_pairwise)), and the implication from multiset.pairwise to set.pairwise ([multiset.pairwise.forall](https://leanprover-community.github.io/mathlib_docs/data/multiset/nodup.html#multiset.pairwise.forall)), both under the condition that the relation is symmetric.